File: resolve_importer/Import Meteors.lua
-- Resolve Meteor Detector - in-Resolve Lua importer
-- Targeted at DaVinci Resolve Studio 21.x.
--
-- Install this file in Resolve's Fusion/Scripts/Utility folder, restart Resolve,
-- then run it from Workspace > Scripts > Import Meteors.
--
-- The script is intentionally self-contained: no external Lua modules are needed.

local MARKER_COLOR = "Pink"
local MARKER_NAME = "Meteor"
local CUSTOM_PREFIX = "resolve-meteor-detector:"

-- ---------------------------------------------------------------------------
-- Minimal JSON decoder (objects, arrays, strings, numbers, booleans and null)
-- ---------------------------------------------------------------------------

local JSON_NULL = {}

local function json_error(text, pos, msg)
    error(string.format("JSON parse error at byte %d: %s", pos, msg))
end

local function skip_ws(text, pos)
    while true do
        local c = text:sub(pos, pos)
        if c == " " or c == "\t" or c == "\r" or c == "\n" then
            pos = pos + 1
        else
            return pos
        end
    end
end

local parse_value

local function parse_string(text, pos)
    if text:sub(pos, pos) ~= '"' then json_error(text, pos, "expected string") end
    pos = pos + 1
    local out = {}
    while pos <= #text do
        local c = text:sub(pos, pos)
        if c == '"' then
            return table.concat(out), pos + 1
        elseif c == "\\" then
            local esc = text:sub(pos + 1, pos + 1)
            if esc == '"' or esc == "\\" or esc == "/" then
                out[#out + 1] = esc
                pos = pos + 2
            elseif esc == "b" then out[#out + 1] = "\b"; pos = pos + 2
            elseif esc == "f" then out[#out + 1] = "\f"; pos = pos + 2
            elseif esc == "n" then out[#out + 1] = "\n"; pos = pos + 2
            elseif esc == "r" then out[#out + 1] = "\r"; pos = pos + 2
            elseif esc == "t" then out[#out + 1] = "\t"; pos = pos + 2
            elseif esc == "u" then
                local hex = text:sub(pos + 2, pos + 5)
                if #hex ~= 4 or not hex:match("^[0-9a-fA-F]+$") then
                    json_error(text, pos, "invalid unicode escape")
                end
                local cp = tonumber(hex, 16)
                -- Handle UTF-16 surrogate pairs.
                if cp >= 0xD800 and cp <= 0xDBFF and text:sub(pos + 6, pos + 7) == "\\u" then
                    local hex2 = text:sub(pos + 8, pos + 11)
                    local low = tonumber(hex2, 16)
                    if low and low >= 0xDC00 and low <= 0xDFFF then
                        cp = 0x10000 + (cp - 0xD800) * 0x400 + (low - 0xDC00)
                        pos = pos + 6
                    end
                end
                if cp <= 0x7F then
                    out[#out + 1] = string.char(cp)
                elseif cp <= 0x7FF then
                    out[#out + 1] = string.char(
                        0xC0 + math.floor(cp / 0x40),
                        0x80 + (cp % 0x40)
                    )
                elseif cp <= 0xFFFF then
                    out[#out + 1] = string.char(
                        0xE0 + math.floor(cp / 0x1000),
                        0x80 + (math.floor(cp / 0x40) % 0x40),
                        0x80 + (cp % 0x40)
                    )
                else
                    out[#out + 1] = string.char(
                        0xF0 + math.floor(cp / 0x40000),
                        0x80 + (math.floor(cp / 0x1000) % 0x40),
                        0x80 + (math.floor(cp / 0x40) % 0x40),
                        0x80 + (cp % 0x40)
                    )
                end
                pos = pos + 6
            else
                json_error(text, pos, "invalid escape sequence")
            end
        else
            if c:byte() < 0x20 then json_error(text, pos, "control character in string") end
            out[#out + 1] = c
            pos = pos + 1
        end
    end
    json_error(text, pos, "unterminated string")
end

local function parse_number(text, pos)
    local start = pos
    local c = text:sub(pos, pos)
    if c == "-" then pos = pos + 1 end
    if text:sub(pos, pos) == "0" then
        pos = pos + 1
    else
        if not text:sub(pos, pos):match("%d") then json_error(text, pos, "invalid number") end
        while text:sub(pos, pos):match("%d") do pos = pos + 1 end
    end
    if text:sub(pos, pos) == "." then
        pos = pos + 1
        if not text:sub(pos, pos):match("%d") then json_error(text, pos, "invalid fraction") end
        while text:sub(pos, pos):match("%d") do pos = pos + 1 end
    end
    c = text:sub(pos, pos)
    if c == "e" or c == "E" then
        pos = pos + 1
        c = text:sub(pos, pos)
        if c == "+" or c == "-" then pos = pos + 1 end
        if not text:sub(pos, pos):match("%d") then json_error(text, pos, "invalid exponent") end
        while text:sub(pos, pos):match("%d") do pos = pos + 1 end
    end
    local n = tonumber(text:sub(start, pos - 1))
    if n == nil then json_error(text, start, "invalid number") end
    return n, pos
end

local function parse_array(text, pos)
    local arr = {}
    pos = skip_ws(text, pos + 1)
    if text:sub(pos, pos) == "]" then return arr, pos + 1 end
    local i = 1
    while true do
        local value
        value, pos = parse_value(text, pos)
        arr[i] = value
        i = i + 1
        pos = skip_ws(text, pos)
        local c = text:sub(pos, pos)
        if c == "]" then return arr, pos + 1 end
        if c ~= "," then json_error(text, pos, "expected ',' or ']'") end
        pos = skip_ws(text, pos + 1)
    end
end

local function parse_object(text, pos)
    local obj = {}
    pos = skip_ws(text, pos + 1)
    if text:sub(pos, pos) == "}" then return obj, pos + 1 end
    while true do
        if text:sub(pos, pos) ~= '"' then json_error(text, pos, "expected object key") end
        local key
        key, pos = parse_string(text, pos)
        pos = skip_ws(text, pos)
        if text:sub(pos, pos) ~= ":" then json_error(text, pos, "expected ':'") end
        pos = skip_ws(text, pos + 1)
        local value
        value, pos = parse_value(text, pos)
        obj[key] = value
        pos = skip_ws(text, pos)
        local c = text:sub(pos, pos)
        if c == "}" then return obj, pos + 1 end
        if c ~= "," then json_error(text, pos, "expected ',' or '}'") end
        pos = skip_ws(text, pos + 1)
    end
end

parse_value = function(text, pos)
    pos = skip_ws(text, pos)
    local c = text:sub(pos, pos)
    if c == '"' then return parse_string(text, pos) end
    if c == "{" then return parse_object(text, pos) end
    if c == "[" then return parse_array(text, pos) end
    if c == "-" or c:match("%d") then return parse_number(text, pos) end
    if text:sub(pos, pos + 3) == "true" then return true, pos + 4 end
    if text:sub(pos, pos + 4) == "false" then return false, pos + 5 end
    if text:sub(pos, pos + 3) == "null" then return JSON_NULL, pos + 4 end
    json_error(text, pos, "unexpected token")
end

local function json_decode(text)
    local value, pos = parse_value(text, 1)
    pos = skip_ws(text, pos)
    if pos <= #text then json_error(text, pos, "trailing data") end
    return value
end

-- ---------------------------------------------------------------------------
-- Helpers
-- ---------------------------------------------------------------------------

local function basename(path)
    path = tostring(path or "")
    path = path:gsub("\\", "/")
    return path:match("([^/]+)$") or path
end

local function norm_name(path)
    return string.lower(basename(path))
end

local function file_exists(path)
    local f = io.open(path, "rb")
    if f then f:close(); return true end
    return false
end

local function read_all(path)
    local f, err = io.open(path, "rb")
    if not f then return nil, err end
    local data = f:read("*a")
    f:close()
    return data
end

local function home_dir()
    return os.getenv("HOME") or os.getenv("USERPROFILE") or ""
end

local function get_fusion()
    if fusion ~= nil then return fusion end
    if fu ~= nil then return fu end
    if type(Fusion) == "function" then
        local ok, result = pcall(Fusion)
        if ok then return result end
    end
    return nil
end

local function get_resolve()
    if type(Resolve) == "function" then
        local ok, result = pcall(Resolve)
        if ok and result then return result end
    end
    local f = get_fusion()
    if f and f.GetResolve then
        local ok, result = pcall(function() return f:GetResolve() end)
        if ok and result then return result end
    end
    if app and app.GetResolve then
        local ok, result = pcall(function() return app:GetResolve() end)
        if ok and result then return result end
    end
    return nil
end

local function request_json_path()
    local f = get_fusion()

    -- Fusion/Resolve builds commonly expose RequestFile on the Fusion application.
    if f and f.RequestFile then
        local ok, result = pcall(function()
            return f:RequestFile("", "JSON files (*.json)|*.json", { FReqB_SeqGather = false })
        end)
        if ok and result and tostring(result) ~= "" then return tostring(result) end
    end

    -- AskUser's FileBrowse control is part of the embedded Fusion scripting API.
    -- It needs a current Fusion composition, so this is a fallback rather than the
    -- primary picker when running from the Edit page.
    if f and f.GetCurrentComp then
        local ok_comp, comp = pcall(function() return f:GetCurrentComp() end)
        if ok_comp and comp and comp.AskUser then
            local ok, result = pcall(function()
                return comp:AskUser("Import Meteor Markers", {
                    { "JSONFile", "FileBrowse", Name = "Meteor results JSON" }
                })
            end)
            if ok and result and result.JSONFile and tostring(result.JSONFile) ~= "" then
                return tostring(result.JSONFile)
            end
        end
    end

    -- Non-interactive fallback for unusual Resolve/Fusion builds.
    local env = os.getenv("METEOR_JSON")
    if env and env ~= "" and file_exists(env) then return env end
    local home = home_dir()
    if home ~= "" then
        local fallback = home .. "/meteors.json"
        if file_exists(fallback) then return fallback end
    end
    return nil
end

local function clip_filename(item)
    local ok, mpi = pcall(function() return item:GetMediaPoolItem() end)
    if not ok or not mpi then return "" end

    local props = {}
    pcall(function() props = mpi:GetClipProperty() or {} end)
    local keys = { "File Path", "FilePath", "Filename", "File Name" }
    for _, key in ipairs(keys) do
        if props[key] and tostring(props[key]) ~= "" then
            return basename(props[key])
        end
    end

    local name = nil
    pcall(function() name = mpi:GetName() end)
    if name and tostring(name) ~= "" then return tostring(name) end
    pcall(function() name = item:GetName() end)
    return tostring(name or "")
end

local function source_range(item)
    local left, duration
    local ok1 = pcall(function() left = tonumber(item:GetLeftOffset()) end)
    local ok2 = pcall(function() duration = tonumber(item:GetDuration()) end)
    if not ok1 or not ok2 or not left or not duration or duration <= 0 then return nil, nil end
    left = math.floor(left + 0.5)
    duration = math.floor(duration + 0.5)
    return left, left + duration - 1
end

local function existing_custom_data(item)
    local found = {}
    local markers = {}
    pcall(function() markers = item:GetMarkers() or {} end)
    for _, info in pairs(markers) do
        if type(info) == "table" then
            local custom = info.customData or info.customdata
            if custom and tostring(custom) ~= "" then found[tostring(custom)] = true end
        end
    end
    return found
end

local function join_sorted_keys(set)
    local list = {}
    for key, _ in pairs(set) do list[#list + 1] = key end
    table.sort(list)
    return table.concat(list, ", ")
end

local function show_message(title, body)
    print(title)
    print(body)
    local f = get_fusion()
    if f and f.GetCurrentComp then
        local ok_comp, comp = pcall(function() return f:GetCurrentComp() end)
        if ok_comp and comp and comp.AskUser then
            pcall(function()
                comp:AskUser(title, {
                    { "Message", "Text", Default = body, ReadOnly = true, Lines = 12, Wrap = true }
                })
            end)
        end
    end
end

-- ---------------------------------------------------------------------------
-- Import
-- ---------------------------------------------------------------------------

local function main()
    local json_path = request_json_path()
    if not json_path then
        show_message(
            "Meteor Import",
            "No JSON file was selected. If your Resolve build does not show a file picker, set METEOR_JSON or place meteors.json in your home directory."
        )
        return
    end

    local raw, err = read_all(json_path)
    if not raw then error("Could not read JSON file: " .. tostring(err)) end
    local data = json_decode(raw)
    if type(data) ~= "table" or data.format ~= "resolve-meteor-detector" then
        error("Not a resolve-meteor-detector JSON file")
    end

    local detections = {}
    local duplicates = {}
    for _, fdata in ipairs(data.files or {}) do
        local name = norm_name(fdata.filename or "")
        if name ~= "" then
            if detections[name] then duplicates[name] = true end
            detections[name] = fdata
        end
    end
    if next(duplicates) ~= nil then
        error("Duplicate source filenames in JSON are ambiguous: " .. join_sorted_keys(duplicates))
    end

    local resolve = get_resolve()
    if not resolve then error("Could not access the DaVinci Resolve scripting object") end
    local pm = resolve:GetProjectManager()
    local project = pm and pm:GetCurrentProject() or nil
    local timeline = project and project:GetCurrentTimeline() or nil
    if not timeline then error("No current timeline is open") end

    local added, skipped_trim, already, matched_items = 0, 0, 0, 0
    local unmatched = {}
    for name, _ in pairs(detections) do unmatched[name] = true end

    local track_count = tonumber(timeline:GetTrackCount("video") or 0) or 0
    for track = 1, track_count do
        local items = timeline:GetItemListInTrack("video", track) or {}
        for _, item in pairs(items) do
            local filename = norm_name(clip_filename(item))
            local fdata = detections[filename]
            if filename ~= "" and fdata then
                matched_items = matched_items + 1
                unmatched[filename] = nil
                local existing = existing_custom_data(item)
                local src_first, src_last = source_range(item)

                for _, event in ipairs(fdata.events or {}) do
                    local source_frame = tonumber(event.peak_frame)
                    if source_frame then
                        source_frame = math.floor(source_frame + 0.5)
                        if src_first and (source_frame < src_first or source_frame > src_last) then
                            skipped_trim = skipped_trim + 1
                        else
                            local event_id = tostring(event.id or (basename(filename):gsub("%.[^.]+$", "") .. "-" .. source_frame))
                            local custom = CUSTOM_PREFIX .. event_id
                            if existing[custom] then
                                already = already + 1
                            else
                                local note = string.format(
                                    "Detected meteor\nSource: %s\nFrames: %s-%s\nPeak frame: %d\nConfidence: %s",
                                    tostring(fdata.filename or filename),
                                    tostring(event.start_frame or "?"),
                                    tostring(event.end_frame or "?"),
                                    source_frame,
                                    tostring(event.confidence or "n/a")
                                )
                                local ok = item:AddMarker(source_frame, MARKER_COLOR, MARKER_NAME, note, 1, custom)
                                if ok then
                                    added = added + 1
                                    existing[custom] = true
                                else
                                    print(string.format(
                                        "WARNING: Resolve rejected marker %s on %s at source frame %d",
                                        event_id, filename, source_frame
                                    ))
                                end
                            end
                        end
                    end
                end
            end
        end
    end

    local unmatched_count = 0
    for _ in pairs(unmatched) do unmatched_count = unmatched_count + 1 end
    local timeline_name = "(unnamed)"
    pcall(function() timeline_name = tostring(timeline:GetName()) end)

    local summary = string.format(
        "Timeline: %s\nJSON: %s\n\nMatched clip instances: %d\nAdded Pink clip markers: %d\nAlready present: %d\nSkipped (trimmed out): %d\nJSON files not on timeline: %d",
        timeline_name, json_path, matched_items, added, already, skipped_trim, unmatched_count
    )
    if unmatched_count > 0 then
        summary = summary .. "\n\nUnmatched files:\n" .. join_sorted_keys(unmatched)
    end
    show_message("Meteor Import Complete", summary)
end

local ok, err = xpcall(main, debug.traceback)
if not ok then
    show_message("Meteor Import Error", tostring(err))
end

File: resolve_importer/import_meteors.py
#!/usr/bin/env python3
"""Import meteor detections as Pink *clip markers* in the current Resolve timeline.

Designed for DaVinci Resolve Studio 21.0.4 on Linux, while using only longstanding
scripting calls. Run externally with Resolve open:

    python import_meteors.py /path/to/meteors.json

It can also be installed as a Resolve Workspace script. In that mode, set
METEOR_JSON=/path/to/meteors.json or put meteors.json in your home directory.
"""
from __future__ import annotations

import json
import os
import sys
from pathlib import Path
from typing import Any

MARKER_COLOR = "Pink"
MARKER_NAME = "Meteor"
CUSTOM_PREFIX = "resolve-meteor-detector:"


def _import_resolve_module():
    try:
        import DaVinciResolveScript as dvr_script  # type: ignore
        return dvr_script
    except ImportError:
        candidates = [
            os.environ.get("RESOLVE_SCRIPT_API"),
            "/opt/resolve/Developer/Scripting",
            "/opt/resolve/Developer/Scripting/Modules",
        ]
        for base in candidates:
            if not base:
                continue
            p = Path(base)
            module_dir = p / "Modules" if (p / "Modules").is_dir() else p
            if module_dir.is_dir() and str(module_dir) not in sys.path:
                sys.path.insert(0, str(module_dir))
        try:
            import DaVinciResolveScript as dvr_script  # type: ignore
            return dvr_script
        except ImportError as exc:
            raise RuntimeError(
                "Could not import DaVinciResolveScript. On a normal Linux install, set:\n"
                "  export RESOLVE_SCRIPT_API=/opt/resolve/Developer/Scripting\n"
                "  export RESOLVE_SCRIPT_LIB=/opt/resolve/libs/Fusion/fusionscript.so\n"
                "  export PYTHONPATH=\"$PYTHONPATH:$RESOLVE_SCRIPT_API/Modules\""
            ) from exc


def _json_path() -> Path:
    if len(sys.argv) >= 2 and not sys.argv[1].startswith("-"):
        return Path(sys.argv[1]).expanduser().resolve()
    env = os.environ.get("METEOR_JSON")
    if env:
        return Path(env).expanduser().resolve()
    return Path.home() / "meteors.json"


def _norm_name(s: str) -> str:
    return Path(s or "").name.casefold()


def _clip_filename(item) -> str:
    mpi = item.GetMediaPoolItem()
    if not mpi:
        return ""
    # File Path is the most reliable property when available. Fall back to clip name.
    try:
        props = mpi.GetClipProperty() or {}
    except Exception:
        props = {}
    for key in ("File Path", "FilePath", "Filename", "File Name"):
        if props.get(key):
            return Path(str(props[key])).name
    try:
        name = mpi.GetName()
        if name:
            return str(name)
    except Exception:
        pass
    try:
        return str(item.GetName() or "")
    except Exception:
        return ""


def _existing_custom_data(item) -> set[str]:
    found: set[str] = set()
    try:
        markers = item.GetMarkers() or {}
    except Exception:
        markers = {}
    for info in markers.values():
        if isinstance(info, dict):
            c = info.get("customData") or info.get("customdata")
            if c:
                found.add(str(c))
    return found


def _source_range(item) -> tuple[int, int] | None:
    """Return inclusive source-relative frame range represented by this timeline item.

    Resolve TimelineItem marker frame IDs are source-relative. GetLeftOffset() is the
    source offset at the visible beginning of the edit. This also lets us skip detections
    that were trimmed out of a timeline instance.
    """
    try:
        left = int(round(float(item.GetLeftOffset())))
        duration = int(round(float(item.GetDuration())))
        if duration <= 0:
            return None
        return left, left + duration - 1
    except Exception:
        return None


def _frame_mode() -> str:
    mode = os.environ.get("METEOR_FRAME_MODE", "source").strip().lower()
    for arg in sys.argv[2:]:
        if arg.startswith("--frame-mode="):
            mode = arg.split("=", 1)[1].strip().lower()
    if mode not in {"source", "clip-relative"}:
        raise RuntimeError("METEOR_FRAME_MODE/--frame-mode must be source or clip-relative")
    return mode


def main() -> int:
    jpath = _json_path()
    frame_mode = _frame_mode()
    if not jpath.exists():
        print(f"Meteor JSON not found: {jpath}")
        print("Pass it as argv[1], set METEOR_JSON, or copy it to ~/meteors.json")
        return 2
    with jpath.open("r", encoding="utf-8") as f:
        data = json.load(f)
    if data.get("format") != "resolve-meteor-detector":
        raise RuntimeError("Not a resolve-meteor-detector JSON file")

    detections: dict[str, dict[str, Any]] = {}
    duplicates: set[str] = set()
    for fdata in data.get("files", []):
        name = _norm_name(str(fdata.get("filename", "")))
        if not name:
            continue
        if name in detections:
            duplicates.add(name)
        detections[name] = fdata
    if duplicates:
        raise RuntimeError("Duplicate source filenames in JSON are ambiguous: " + ", ".join(sorted(duplicates)))

    dvr_script = _import_resolve_module()
    resolve = dvr_script.scriptapp("Resolve")
    if not resolve:
        raise RuntimeError("Could not connect to Resolve. Make sure Resolve is running and external scripting is allowed.")
    pm = resolve.GetProjectManager()
    project = pm.GetCurrentProject() if pm else None
    timeline = project.GetCurrentTimeline() if project else None
    if not timeline:
        raise RuntimeError("No current timeline is open")

    added = 0
    skipped_trim = 0
    already = 0
    unmatched: set[str] = set(detections.keys())
    matched_items = 0

    track_count = int(timeline.GetTrackCount("video") or 0)
    for track in range(1, track_count + 1):
        items = timeline.GetItemListInTrack("video", track) or []
        for item in items:
            filename = _norm_name(_clip_filename(item))
            if not filename or filename not in detections:
                continue
            matched_items += 1
            unmatched.discard(filename)
            fdata = detections[filename]
            existing = _existing_custom_data(item)
            src_range = _source_range(item)

            for event in fdata.get("events", []):
                source_frame = int(event["peak_frame"])
                if src_range is not None and not (src_range[0] <= source_frame <= src_range[1]):
                    skipped_trim += 1
                    continue
                event_id = str(event.get("id") or f"{Path(filename).stem}-{source_frame}")
                custom = CUSTOM_PREFIX + event_id
                if custom in existing:
                    already += 1
                    continue
                note = (
                    f"Detected meteor\n"
                    f"Source: {fdata.get('filename', filename)}\n"
                    f"Frames: {event.get('start_frame')}–{event.get('end_frame')}\n"
                    f"Peak frame: {source_frame}\n"
                    f"Confidence: {event.get('confidence', 'n/a')}"
                )
                # Resolve versions/workflows can be validated with either mapping. The default
                # is source-relative; clip-relative is available as an explicit fallback.
                marker_frame = source_frame
                if frame_mode == "clip-relative" and src_range is not None:
                    marker_frame = source_frame - src_range[0]
                ok = item.AddMarker(marker_frame, MARKER_COLOR, MARKER_NAME, note, 1, custom)
                if ok:
                    added += 1
                    existing.add(custom)
                else:
                    print(f"WARNING: Resolve rejected marker {event_id} on {filename} at source frame {source_frame}")

    print(f"Timeline: {timeline.GetName()}")
    print(f"Marker frame mode: {frame_mode}")
    print(f"Matched timeline clip instances: {matched_items}")
    print(f"Added Pink clip markers: {added}")
    print(f"Already present: {already}")
    print(f"Skipped because detection is outside a trimmed edit: {skipped_trim}")
    if unmatched:
        print("JSON files not present on this timeline:")
        for name in sorted(unmatched):
            print(f"  {name}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
